src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -55,28 +55,60 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
     //    an invalid state for the `ActionMap`
     Windows::Foundation::Collections::IMapView<hstring, Model::Command> ActionMap::NameMap()
     {
-        if (_NameMapCache)
+        if (!_NameMapCache)
         {
-            return _NameMapCache.GetView();
+            // populate _NameMapCache
+            std::unordered_map<hstring, Model::Command> nameMap{};
+            std::set<InternalActionID> visitedActionIDs{ ActionHash()(make<implementation::ActionAndArgs>()) };
+            _PopulateNameMapWithNestedCommands(nameMap);
+            _PopulateNameMapWithStandardCommands(nameMap, visitedActionIDs);
+
+            _NameMapCache = single_threaded_map<hstring, Model::Command>(std::move(nameMap));
         }
+        return _NameMapCache.GetView();
+    }
 
-        std::unordered_map<hstring, Model::Command> nameMap{};
-        std::set<size_t> visitedActionIDs{ ActionHash()(make<implementation::ActionAndArgs>()) };
-        _PopulateNameMap(nameMap, visitedActionIDs);
+    // Method Description:
+    // - Populates the provided nameMap with all of our nested commands and our parents nested commands
+    // - Performs a top-down approach by going to the root first, then recursively adding the nested commands layer-by-layer
+    // Arguments:
+    // - nameMap: the nameMap we're populating. This maps the name (hstring) of a command to the command itself.
+    void ActionMap::_PopulateNameMapWithNestedCommands(std::unordered_map<hstring, Model::Command>& nameMap) const
+    {
+        // Update NameMap with our parents.
+        // Starting with this means we're doing a top-down approach.
+        for (const auto& parent : _parents)
+        {
+            parent->_PopulateNameMapWithNestedCommands(nameMap);
+        }
 
-        _NameMapCache = single_threaded_map<hstring, Model::Command>(std::move(nameMap));
-        return _NameMapCache.GetView();
+        // Add NestedCommands to NameMap _after_ we handle our parents.
+        // This allows us to override whatever our parents tell us.
+        for (const auto& [name, cmd] : _NestedCommands)
+        {
+            if (cmd.HasNestedCommands())
+            {
+                // add a valid cmd
+                nameMap.insert_or_assign(name, cmd);
+            }
+            else
+            {
+                // remove the invalid cmd
+                nameMap.erase(name);
+            }

```

Comment:
This theoretically can't get hit, right? Like, a command should never be placed into `_NestedCommands` without nested commands in the first place, right?

Reply:
Actually, it can! This is that special differentiation between `HasNestedCommands` and `IsNestedCommand`. In `AddAction`, we populate `_NestedCommands` with all of the `IsNestedCommand == true` commands (allowing unbinding). Here, we use `HasNestedCommands` to differentiate between those meant to unbind and actual nested commands.